### PR TITLE
Remove CA1307 suppressions

### PR DESCRIPTION
**Pull request by martinrrm**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1175

Regression? Last working version:

## Description

Remove CA1307 suppressions 

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Review by jeffkl — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
                 dateTime = dt;
             }
 
-            if (string.Equals(culture.Name, "ja-JP"))
+            if (string.Equals(culture.Name, "ja-JP", StringComparison.OrdinalIgnoreCase))
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs]: I know this won't break anything, but I think we should prefer ordinal in places where there's no clear documentation about casing.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs [hunk 2]
```diff
@@ -139,10 +139,10 @@ private async Task AddFileCoreAsync(string path, Action addFile)
             var lockFileFullPath = PackagesConfigLockFileUtility.GetPackagesLockFilePath(ProjectFullPath, GetPropertyValue("NuGetLockFilePath")?.ToString(), ProjectName);
             if (File.Exists(Path.Combine(ProjectFullPath, path))
                 && !fileExistsInProject
-                && !fileName.Equals(ProjectManagement.Constants.PackageReferenceFile)
-                && !fileName.Equals("packages." + ProjectName + ".config")
-                && !fileName.Equals(EnvDteProjectExtensions.WebConfig)
-                && !fileName.Equals(EnvDteProjectExtensions.AppConfig)
+                && !fileName.Equals(ProjectManagement.Constants.PackageReferenceFile, StringComparison.OrdinalIgnoreCase)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs]: Path comparisons should be aware of the FS casing. 

This applies through the whole PR.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs [hunk 3]
```diff
@@ -75,7 +75,7 @@ private bool ResetSolutionSettingsIfNeeded()
             // In the open PM UI scenario (no restore run), there is an asynchronous invocation of this code path. This changes ensures that
             // the synchronous calls that come after the asynchrnous calls don't do duplicate work.
             // That however is not the case for solution close and  same session close -> open events. Those will be on the UI thread.
-            if (_solutionSettings == null || !string.Equals(root, _solutionSettings.Item1))
+            if (_solutionSettings == null || !string.Equals(root, _solutionSettings.Item1, StringComparison.OrdinalIgnoreCase))
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs]: Same as well, Path comparisons should be aware of the FS casing. 
We have a helper method, OsAware or something like that.

**Review by nkolev92 — COMMENTED** [review]
I don't expect it to break anything in most cases, but we should make sure path comparisons are actually FS aware. 

PR naming: It's probably best to explain what the change is instead of just refering the warning code.

Example: "Specify StringComparison for clarity, CA1307"

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs [hunk 4]
```diff
@@ -818,7 +818,7 @@ private static bool IsSolutionOrProjectFile(string fileName)
         /// </summary>
         private static bool IsPackagesConfig(string projectFileName)
         {
-            return string.Equals(projectFileName, Constants.PackageReferenceFile)
+            return string.Equals(projectFileName, Constants.PackageReferenceFile, StringComparison.OrdinalIgnoreCase)
```
Comment (Youssef1313) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs]:
```suggestion
            return string.Equals(projectFileName, Constants.PackageReferenceFile, StringComparison.Ordinal)
```

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs [hunk 5]
```diff
@@ -448,7 +448,7 @@ private static bool IsWindows10(ILogger logger)
         {
             var productName = (string)RegistryKeyUtility.GetValueFromRegistryKey("ProductName", OSVersionRegistryKey, Registry.LocalMachine, logger);
 
-            return productName != null && productName.StartsWith("Windows 10");
+            return productName != null && productName.StartsWith("Windows 10", StringComparison.OrdinalIgnoreCase);
```
Comment (Youssef1313) [inline src/NuGet.Clients/NuGet.CommandLine/Program.cs]:
```suggestion
            return productName != null && productName.StartsWith("Windows 10", StringComparison.Ordinal);
```

File: src/NuGet.Clients/NuGet.Console/WpfConsole/WpfConsoleService.cs [hunk 6]
```diff
@@ -98,7 +98,7 @@ public object GetClassifier(object textBuffer)
                 {
                     IService service = null;
 
-                    var lazyProvider = providers.FirstOrDefault(f => string.Equals(f.Metadata.HostName, console.HostName));
+                    var lazyProvider = providers.FirstOrDefault(f => string.Equals(f.Metadata.HostName, console.HostName, StringComparison.OrdinalIgnoreCase));
```
Comment (Youssef1313) [inline src/NuGet.Clients/NuGet.Console/WpfConsole/WpfConsoleService.cs]:
```suggestion
                    var lazyProvider = providers.FirstOrDefault(f => string.Equals(f.Metadata.HostName, console.HostName, StringComparison.Ordinal));
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs (hunk 1 above)
Comment (Youssef1313) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs]:
```suggestion
            if (string.Equals(culture.Name, "ja-JP", StringComparison.Ordinal))
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/AuthorAndDownloadCount.xaml.cs [hunk 7]
```diff
@@ -101,7 +102,7 @@ private void UpdateControl()
                 var formatString = Resx.Text_Downloads;
                 string begin = string.Empty;
                 string end = string.Empty;
-                var index = formatString.IndexOf("{0}");
+                var index = formatString.IndexOf("{0}", StringComparison.OrdinalIgnoreCase);
```
Comment (Youssef1313) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/AuthorAndDownloadCount.xaml.cs]:
```suggestion
                var index = formatString.IndexOf("{0}", StringComparison.Ordinal);
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs (hunk 2 above)
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs]: Here the comparison in for the file name only, should this also be aware of the OS?

File: build/NuGet.ruleset [hunk 8]
```diff
@@ -10,6 +10,9 @@
   <Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.CodeStyle" RuleNamespace="Microsoft.CodeQuality.Analyzers">
     <Rule Id="IDE0055" Action="Warning" />
   </Rules>
+  <Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.NetAnalyzers" RuleNamespace="Microsoft.CodeAnalysis.CSharp.NetAnalyzers">
+    <Rule Id="CA1309" Action="Error" />
+  </Rules>
```
Comment (Youssef1313) [inline build/NuGet.ruleset]: Not related to the PR, but rulesets are considered deprecated. Consider moving to globalconfigs or editorconfigs

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs (hunk 1 above)
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/DateTimeConverter.cs]: I did a better investigation this time and realized that the methods `.Equals()`, `.EndsWith()`, `.StartsWith()` and `.IndexOf()` when not specifying the StringComparison is going to use `Ordinal` as default. So in order to not change current behavior I'm going to use Ordinal instead

https://learn.microsoft.com/en-us/dotnet/api/system.string.equals?view=net-6.0#system-string-equals(system-string)
https://learn.microsoft.com/en-us/dotnet/api/system.string.endswith?view=net-7.0#system-string-endswith(system-char)
https://learn.microsoft.com/en-us/dotnet/api/system.string.startswith?view=net-7.0#system-string-startswith(system-string)
https://learn.microsoft.com/en-us/dotnet/api/system.string.indexof?view=net-7.0#system-string-indexof(system-string)

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 9]
```diff
@@ -201,7 +201,7 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
                 }
 
                 // Get package version if it already exists in the props file. Returns null if there is no matching package version.
-                ProjectItem packageVersionInProps = project.Items.LastOrDefault(i => i.ItemType == PACKAGE_VERSION_TYPE_TAG && i.EvaluatedInclude.Equals(libraryDependency.Name));
+                ProjectItem packageVersionInProps = project.Items.LastOrDefault(i => i.ItemType == PACKAGE_VERSION_TYPE_TAG && i.EvaluatedInclude.Equals(libraryDependency.Name, StringComparison.Ordinal));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: I think this codepath had a bug. 

package ids are case insensitive, so by switching to ordinalignore case we're actually fixing a bug.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 10]
```diff
@@ -378,7 +378,7 @@ private string AddVersionMetadata(LibraryDependency libraryDependency, ProjectIt
             var packageVersion = libraryDependency.LibraryRange.VersionRange.OriginalString ??
                     libraryDependency.LibraryRange.VersionRange.MinVersion.ToString();
 
-            ProjectMetadataElement versionAttribute = item.Metadata.FirstOrDefault(i => i.Name.Equals("Version"));
+            ProjectMetadataElement versionAttribute = item.Metadata.FirstOrDefault(i => i.Name.Equals("Version", StringComparison.Ordinal));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Similar as the previous one. MSBuild metadata is case insensitive.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 11]
```diff
@@ -434,7 +434,7 @@ internal void UpdatePackageVersion(Project project, ProjectItem packageVersion,
             ProjectItemElement packageVersionItemElement = project.GetItemProvenance(packageVersion).LastOrDefault()?.ItemElement;
 
             // Get the Version attribute on the packageVersionItemElement.
-            ProjectMetadataElement versionAttribute = packageVersionItemElement.Metadata.FirstOrDefault(i => i.Name.Equals("Version"));
+            ProjectMetadataElement versionAttribute = packageVersionItemElement.Metadata.FirstOrDefault(i => i.Name.Equals("Version", StringComparison.Ordinal));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: case insensitive here as well.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 12]
```diff
@@ -746,7 +746,7 @@ private static IEnumerable<InstalledPackageReference> GetPackageReferencesFromTa
             var newProject = new ProjectInstance(projectPath, globalProperties, null);
             newProject.Build(new[] { CollectPackageReferences }, new List<Microsoft.Build.Framework.ILogger> { }, out var targetOutputs);
 
-            return targetOutputs.First(e => e.Key.Equals(CollectPackageReferences)).Value.Items.Select(p =>
+            return targetOutputs.First(e => e.Key.Equals(CollectPackageReferences, StringComparison.Ordinal)).Value.Items.Select(p =>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: MSBuild targets are case insensitive too.

File: src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs [hunk 13]
```diff
@@ -490,7 +490,7 @@ public static string StripLeadingDirectorySeparators(string filename)
             filename = GetPathWithForwardSlashes(filename);
             var currentDirectoryPath = $"./";
 
-            if (filename.StartsWith(currentDirectoryPath))
+            if (filename.StartsWith(currentDirectoryPath, StringComparison.Ordinal))
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs]: This deals with paths, so probably should be fs aware?

**Review by nkolev92 — COMMENTED** [review]
A few small changes. 

I'm not a huge fan how 80% of changes we need to make for the rule are just specifiyng ordinal which just increases the complexity, but a necessary evil I guess.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs [hunk 14]
```diff
@@ -818,7 +818,7 @@ private static bool IsSolutionOrProjectFile(string fileName)
         /// </summary>
         private static bool IsPackagesConfig(string projectFileName)
         {
-            return string.Equals(projectFileName, Constants.PackageReferenceFile)
+            return string.Equals(projectFileName, Constants.PackageReferenceFile, StringComparison.Ordinal)
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs]: This method seems to be comparing paths. I suggest using the helper method to compare string paths based on OS.

```suggestion
            return string.Equals(projectFileName, Constants.PackageReferenceFile, PathUtil.GetStringComparisonBasedOnOS())
```

**Review by dominoFire — DISMISSED** [review]
Partial review. Thank you!

**Review by dominoFire — COMMENTED** [review]
I mean to comment, not 'request changes' optin

File: src/NuGet.Core/NuGet.Frameworks/GlobalSuppressions.cs [hunk 15]
```diff
@@ -28,7 +28,6 @@
 [assembly: SuppressMessage("Build", "CA1062:In externally visible method 'NuGetFramework.NuGetFramework(NuGetFramework framework)', validate parameter 'framework' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.Frameworks.NuGetFramework.#ctor(NuGet.Frameworks.NuGetFramework)")]
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Frameworks/GlobalSuppressions.cs:28]: Discussed offline, remaining CA1307 will be addressed later.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs [hunk 16]
```diff
@@ -40,7 +40,7 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
             }
             //If the given file is a solution, get the list of projects
             //If not, then it's a project, which is put in a list
-            var projectsPaths = Path.GetExtension(listPackageArgs.Path).Equals(".sln") ?
+            var projectsPaths = Path.GetExtension(listPackageArgs.Path).Equals(".sln", StringComparison.Ordinal) ?
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs]: This looks like a path, let's make it File-System aware:

```suggestion
            var projectsPaths = Path.GetExtension(listPackageArgs.Path).Equals(".sln", PathUtil.GetStringComparisonBasedOnOS()) ?
```

File: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs [hunk 17]
```diff
@@ -951,7 +951,7 @@ private IEnumerable<IPackageFile> RemoveDefaultExclusions(ICollection<IPackageFi
                 var fileName = Path.GetFileName(filePath);
 
                 return fileName.EndsWith(".nupkg", StringComparison.OrdinalIgnoreCase)
-                    || (fileName.StartsWith(".") && fileName.IndexOf('.', startIndex: 1) == -1);
+                    || (fileName.StartsWith(".", StringComparison.Ordinal) && fileName.IndexOf('.', startIndex: 1) == -1);
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs]: This looks like a path, but it's just comparing a colon '.'. Let's keep ordinal in both string operations

```suggestion
                    || (fileName.StartsWith(".", StringComparison.Ordinal) && fileName.IndexOf('.', startIndex: 1, StringComparison.Ordinal) == -1);
```

File: src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs [hunk 18]
```diff
@@ -490,7 +490,7 @@ public static string StripLeadingDirectorySeparators(string filename)
             filename = GetPathWithForwardSlashes(filename);
             var currentDirectoryPath = $"./";
 
-            if (filename.StartsWith(currentDirectoryPath))
+            if (filename.StartsWith(currentDirectoryPath, StringComparison.OrdinalIgnoreCase))
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs]: From @nkolev92 [comment](https://github.com/NuGet/NuGet.Client/pull/4888#discussion_r1012330062), we should make this comparison file-system aware:

```suggestion
            if (filename.StartsWith(currentDirectoryPath, PathUtil.GetStringComparisonBasedOnOS()))
```

File: src/NuGet.Core/NuGet.Packaging/PackageFileExtractor.cs [hunk 19]
```diff
@@ -35,7 +36,7 @@ private static HashSet<string> GatherIntellisenseXmlFiles(IEnumerable<string> pa
 
             foreach (var file in refAndLibFiles)
             {
-                if (!file.EndsWith(".xml"))
+                if (!file.EndsWith(".xml", StringComparison.Ordinal))
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Packaging/PackageFileExtractor.cs]: This looks like needs to be file-system aware. But, PackageFileExtractor works on nupkgs, which has standarized filepaths.

I will look at this later to find out if we need FS-awareness or not.

```suggestion
                if (!file.EndsWith(".xml", PathUtil.GetStringComparisonBasedOnOS(()))
```

**Review by dominoFire — COMMENTED** [review]
Good job. Please address comments. Thank you!

**Comment by erdembayar** [comment]
@martinrrm 
Build is failing on CI pipeline.

File: src/NuGet.Core/NuGet.Frameworks/GlobalSuppressions.cs (hunk 15 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.Frameworks/GlobalSuppressions.cs:28]: Follow-up issue: https://github.com/NuGet/Client.Engineering/issues/1977

**Review by dominoFire — APPROVED** [review]

**Pull request merged**